### PR TITLE
Update APIs to use `std::span` instead of `std::vector`. 

### PR DESCRIPTION
**Pull request by greg7mdp**
Also:
- fix issue with `getRandomSeed`
- add versions for `aggregate` functions that take an accessor to avoid vector copies for each call.

### CODE REVIEW

File: bench/chia_bench.cpp [hunk 1]
```diff
@@ -22,24 +22,24 @@ void endStopwatch(string testName,
...
     cout << "Avg: " << now_ms.count() / static_cast<double>(numIters)
-         << " ms" << endl;
+         << " ms" << '\n';
 }
 
-std::vector<uint8_t> getRandomSeed() {
-    uint64_t buf[4];
+std::array<uint8_t, 32> getRandomSeed() {
+    std::array<uint8_t, 32> res;
+    uint64_t *buf = reinterpret_cast<uint64_t *>(res.data());
```
Comment (heifner) [inline bench/chia_bench.cpp:34]: Not 100% sure, but I think this violates strict aliasing rule.

File: bench/chia_bench.cpp [hunk 2]
```diff
@@ -22,24 +22,24 @@ void endStopwatch(string testName,
...
     std::random_device rd;
     std::mt19937_64 gen(rd());
     std::uniform_int_distribution<uint64_t> dis;
     buf[0] = dis(gen);
     buf[1] = dis(gen);
     buf[2] = dis(gen);
     buf[3] = dis(gen);
-    vector<uint8_t> ret(buf, buf + 32);
```
Comment (heifner) [inline bench/chia_bench.cpp:41]: Good find.

File: include/bls12-381/g.hpp [hunk 3]
```diff
@@ -60,7 +60,7 @@ class g1
 
     auto operator<=>(const g1&) const = default;
    
-    static std::optional<g1> multiExp(const std::vector<g1>& points, const std::vector<std::array<uint64_t, 4>>& scalars, std::function<void()> yield = std::function<void()>());
+   static std::optional<g1> multiExp(std::span<const g1> points, std::span<const std::array<uint64_t, 4>> scalars, std::function<void()> yield = std::function<void()>());
```
Comment (heifner) [inline include/bls12-381/g.hpp:63]: Should we remove the unused `yield` ?

File: include/bls12-381/g.hpp [hunk 4]
```diff
@@ -117,7 +117,7 @@ class g2
 
     auto operator<=>(const g2&) const = default;
    
-    static std::optional<g2> multiExp(const std::vector<g2>& points, const std::vector<std::array<uint64_t, 4>>& scalars, std::function<void()> yield = std::function<void()>());
+    static std::optional<g2> multiExp(std::span<const g2> points, std::span<const std::array<uint64_t, 4>> scalars, std::function<void()> yield = std::function<void()>());
```
Comment (heifner) [inline include/bls12-381/g.hpp:120]: Another unused `yield`

File: include/bls12-381/signatures.hpp [hunk 5]
```diff
@@ -100,49 +102,71 @@ g2 derive_child_g2_unhardened(
...
+    std::span<const g1> pubkeys,
+    std::span<const std::vector<uint8_t>> messages,
     const g2& signature,
     const bool checkForDuplicateMessages = false
 );
 
+// f is an accessor function in case we have a vector of object containing public keys
+// pred can be use to aggregate public keys specified in a bit mask for example
+template<class T, class F, class PRED>
+g1 aggregate_public_keys(std::span<const T> pks, F &&f, PRED &&pred) {
```
Comment (heifner) [inline include/bls12-381/signatures.hpp:151]: I think we should default `pred` to `[]{return true;}` or provide a version of function that doesn't take a `pred`.
Also `F&& f` & `Pred&& pred`
Also please add to the comment the expected function signature of `f` and `pred`

File: include/bls12-381/signatures.hpp [hunk 6]
```diff
@@ -100,49 +102,71 @@ g2 derive_child_g2_unhardened(
...
+    for (size_t i=0; i<pks.size(); ++i)
+        if (std::forward<PRED>(pred)(pks[i], i))
+            agg_pk = agg_pk.add(std::forward<F>(f)(pks[i]));
+    return agg_pk;
+}
+
+// f is an accessor function in case we have a vector of object containing signatures
+// pred can be use to aggregate signatures specified in a bit mask for example
+template<class T, class F, class PRED>
+g2 aggregate_signatures(std::span<const T> sigs, F &&f, PRED &&pred) {
```
Comment (heifner) [inline include/bls12-381/signatures.hpp:162]: I think we should default `pred` to `[]{return true;}` or provide a version of function that doesn't take a `pred`.
Also `F&& f` & `Pred&& pred`
Also please add to the comment the expected function signature of `f` and `pred`

**Review by heifner — CHANGES_REQUESTED** [review]

File: include/bls12-381/g.hpp (hunk 3 above)
Comment (greg7mdp) [reply include/bls12-381/g.hpp:63]: Well, it is referenced in the implementation, but I agree it doesn't seem very useful. I don't really know enough about why it was added in the first place to have an opinion on this, can we leave this decision for later?

Comment (heifner) [reply include/bls12-381/g.hpp:63]: Yes, certainly.

**Review by heifner — APPROVED** [review]

File: include/bls12-381/signatures.hpp (hunk 5 above)
Comment (greg7mdp) [reply include/bls12-381/signatures.hpp]: Done. I added some constraints on `F` and `PRED`, and also added a simple test for these functions.

**Review by mschoenebeck — APPROVED** [review]

**Pull request merged**